kura/org.eclipse.kura.driver.eddystone.provider/src/main/java/org/eclipse/kura/internal/driver/eddystone/EddystoneDriver.java

```diff
@@ -0,0 +1,319 @@
+            this.bluetoothLeEddystoneService.deleteBeaconScanner(this.bluetoothLeEddystoneScanner);
+        }
+    }
+
+    @Override
+    public PreparedRead prepareRead(List<ChannelRecord> channelRecords) {
+        logger.warn("Read operation not supported");
+        return new EmptyPreparedRead();
+    }
+
+    @Override
+    public void onBeaconsReceived(BluetoothLeEddystone eddystone) {
+
+        for (EddystoneListener eddystoneListener : this.eddystoneListeners) {
+            if (eddystoneListener.getEddystoneFrameType()
+                    .equals(EddystoneFrameType.valueOf(eddystone.getFrameType()))) {
+                ChannelRecord record = ChannelRecord.createReadRecord(eddystoneListener.getChannelName(),
+                        DataType.STRING);
+
+                record.setValue(TypedValues.newStringValue(eddystoneToString(eddystone)));
+                record.setChannelStatus(new ChannelStatus(SUCCESS));
+                record.setTimestamp(System.currentTimeMillis());
+                eddystoneListener.getListener().onChannelEvent(new ChannelEvent(record));
+            }
+        }
+
+    }
+
+    private String eddystoneToString(BluetoothLeEddystone eddystone) {
+        StringBuilder sb = new StringBuilder();
+        sb.append(eddystone.getFrameType());
+        sb.append(";");
+        if ("UID".equals(eddystone.getFrameType())) {
+            sb.append(bytesArrayToHexString(eddystone.getNamespace()));
+            sb.append(";");
+            sb.append(bytesArrayToHexString(eddystone.getInstance()));
+            sb.append(";");
+        } else if ("URL".equals(eddystone.getFrameType())) {
+            sb.append(eddystone.getUrl());
+            sb.append(";");
+        }
+        sb.append((int) eddystone.getTxPower());
+        sb.append(";");
+        sb.append(eddystone.getRssi());
+        sb.append(";");
+        sb.append(calculateDistance(eddystone.getRssi(), eddystone.getTxPower()));
+
+        return sb.toString();
+    }
+
+    private double calculateDistance(int rssi, int txpower) {
+
+        int ratioDB = txpower - rssi;
+        double ratioLinear = Math.pow(10, (double) ratioDB / 10);
+        return Math.sqrt(ratioLinear);
+    }
+
+    private static String bytesArrayToHexString(byte[] bytes) {
+        StringBuilder sb = new StringBuilder();
+        for (byte b : bytes) {
+            sb.append(String.format("%02X", b));
+        }
+        return sb.toString();
+    }
+
+    private class EmptyPreparedRead implements PreparedRead {
+
+        @Override
+        public void close() throws Exception {
+            // Not needed
+        }
+
+        @Override
+        public List<ChannelRecord> execute() throws ConnectionException, KuraException {
+            logger.warn("Read operation not supported");
+            return new ArrayList<>();
+        }
+
+        @Override
+        public List<ChannelRecord> getChannelRecords() {
+            return new ArrayList<>();
```

Comment:
You can also return `Collections.emptyList()`

Reply:
👍